src/request.rs

```diff
@@ -14,28 +14,113 @@ use super::SerdeValue;
 
 pub type Result<T> = std::result::Result<T, Error>;
 
+/// Parsed result of a request url with handy inspection methods.
 #[derive(Clone)]
-struct ParsedUrl(std::result::Result<Url, url::ParseError>);
+pub struct ParsedUrl {
+    /// Url parse result kept as Result since we want to defer the surfacing of it
+    /// until we do call()/send()/etc.
+    result: std::result::Result<Url, url::ParseError>,
+    origin: Option<String>,
+}
 
 impl fmt::Display for ParsedUrl {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if let Ok(url) = &self.0 {
+        if let Ok(url) = &self.result {
             write!(f, "{}", url.as_str())
         } else {
-            write!(f, "{:?}", self.0)
+            write!(f, "{:?}", self.result)
         }
     }
 }
 
 impl From<String> for ParsedUrl {
     fn from(s: String) -> Self {
-        ParsedUrl(s.parse())
+        let result = s.parse();
+        let origin = if result.is_err() { Some(s) } else { None };
+        ParsedUrl::new(result, origin)
     }
 }
 
 impl From<Url> for ParsedUrl {
     fn from(url: Url) -> Self {
-        ParsedUrl(Ok(url))
+        ParsedUrl::new(Ok(url), None)
+    }
+}
+
+impl ParsedUrl {
+    fn new(mut result: std::result::Result<Url, ParseError>, origin: Option<String>) -> Self {
```

Comment:
I think `mut` is unnecessary here.

Reply:
Given the below changing of `result` if host is empty, it is required.